### ISSUE TITLE
AdamWOptimizer doesn't work with MirroredStrategy and tf.get_variable

### ISSUE THREAD

**Issue by ytdu** [tensorflow/tensorflow#25066]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 16.04
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 1.12.0
- Python version: 3.5.2
- CUDA/cuDNN version: 9.0.176 / 7.4.1
- GPU model and memory: TITAN X, 12189MiB

**Describe the current behavior**
`tf.contrib.opt.AdamWOptimizer` does not do weight decay when using `MirroredStrategy` along with `tf.get_variable`
**Describe the expected behavior**
`tf.contrib.opt.AdamWOptimizer` should always do weight decay, regardless of how variables are created.

**Code to reproduce the issue**
```
import tensorflow as tf

flags = tf.flags
FLAGS = flags.FLAGS
flags.DEFINE_bool('use_get_variable', False, 'Whether to use tf.get_variable to create variables')


def model_fn(features, labels, mode, params):
    if FLAGS.use_get_variable:
        x = tf.get_variable('x', initializer=[1.])
    else:
        x = tf.Variable([1.], name='x')
    loss = tf.reduce_sum(x * x)  # the original loss is 14.

    # here we let learning rate be 0 so that the updates come only from weight decay.
    # after 1 step of weight decay, the final loss should be 0.81.
    optimizer = tf.contrib.opt.AdamWOptimizer(learning_rate=0.0, weight_decay=0.1)

    tvars = tf.trainable_variables()
    tf.logging.info('All variables: {}'.format(tvars))
    grads = tf.gradients(loss, tvars)
    global_step = tf.train.get_or_create_global_step()
    train_op = optimizer.apply_gradients(zip(grads, tvars), global_step=global_step, decay_var_list=tvars)
    output_spec = tf.contrib.tpu.TPUEstimatorSpec(mode=mode, loss=loss, train_op=train_op)
    return output_spec


def input_fn(params):
    return tf.data.Dataset.from_tensors([0.]).repeat()


def main(_):
    tf.logging.set_verbosity(tf.logging.INFO)

    distribution = tf.contrib.distribute.MirroredStrategy()
    run_config = tf.contrib.tpu.RunConfig(train_distribute=distribution)
    estimator = tf.contrib.tpu.TPUEstimator(use_tpu=False, model_fn=model_fn, config=run_config)
    # here max_steps=2 so that the final loss printed is after 1 step update.
    estimator.train(input_fn=input_fn, max_steps=2)


if __name__ == "__main__":
    tf.app.run()
```

**Other info / logs**
I ran the above script with 0 or 1 GPU, both producing:
`python test.py --use_get_variable=False`
![image](https://user-images.githubusercontent.com/7380587/51468725-a1efbd80-1daa-11e9-9328-367769a8c19f.png)

`python test.py --use_get_variable=True`
![image](https://user-images.githubusercontent.com/7380587/51468786-be8bf580-1daa-11e9-8958-d30f1c3dcf19.png)

**Comment by msymp** [comment]
@ytdu , this is an excellent and targeted bug you have detected:
**_tf.contrib.opt.AdamWOptimizer_ does not do weight decay when using _MirroredStrategy_ along with _tf.get_variable_**
@karmel and @josh11b , please advise who can help with this bug. Thanks.

**Comment by josh11b** [comment]
I don't know about this specific issue, but in general we have not supported contrib libraries with tf.distribute.Strategy since they are going away with 2.0.

**Comment by josh11b** [comment]
In general I don't know what happens when you combine TPUEstimator with MirroredStrategy. I'd expect you to use one or the other but not both.

**Comment by ytdu** [comment]
Thanks for the reply. In fact I've gone through some of the related source code, so I know something about why this happens.

`tf.contrib.opt.AdamWOptimizer` (more specifically, `tf.contrib.opt.DecoupledWeightDecayExtension`) maintains a `decay_var_list`, the elements in which are references of variables that are going to be decayed. `MirroredStrategy` somehow wraps all the variables created by `tf.get_variable` into `MirroredVariable`s, which breaks the connection between `decay_var_list` and the variables. Therefore `AdamWOptimizer` does not know these `MirroredVariable`s need decaying.

It would be great that this kind of `MirroredVariable` wrapper related bugs will be eliminated in 2.0

**Comment by guptapriya** [comment]
I don't think this bug will automatically get fixed in 2.0. You're right that the decay_var_list is being stored as the list of mirrored variables, but later on when applying the op, it tries to apply it on the underlying op and that fails. So there needs to be some change in the decay optimizers to handle this. 

Like Josh said, we are not prioritizing bugs with contrib optimizers. But if you have a fix in mind, feel free to send a PR. 

@facaiy - you had expressed interest in the opt module, do you know if there is any progress on this optimizer/is that part of the plan? (https://github.com/tensorflow/community/pull/18/files#r218985470)

**Comment by facaiy** [comment]
tensorflow/addons is planning to migrate AdamWOptimizer: https://github.com/tensorflow/addons/issues/24 . If anyone is interested, welcome to make a contribution. cc @seanpmorgan

**Comment by tensorflowbutler** [comment]
Hi There,

 We are checking to see if you still need help on this issue, as you are using an older version of tensorflow(1.x) which is officially considered as end of life. We recommend that you upgrade to 2.4 or later version and let us know if the issue still persists in newer versions. 

 This issue will be closed automatically 7 days from now. If you still need help with this issue, Please open a new issue for any help you need against 2.x, and we will get you the right help.
